File: apps/desktop/src/cbase/serializer.ts
/**
 * Serialize a LensDefinition back to .cbase YAML format.
 * Used to persist UI-driven view changes (column visibility, sort) to disk.
 */

import { stringify as stringifyYaml } from "yaml";
import type {
  LensDefinition,
  LensFilter,
  LensFilterCondition,
  LensProperty,
  LensSort,
  LensView,
} from "./types";

function serializeProperty(property: LensProperty): Record<string, unknown> {
  const out: Record<string, unknown> = {
    key: property.key,
    type: property.type,
  };
  if (property.label != null) out.label = property.label;
  if (property.required != null) out.required = property.required;
  if (property.default !== undefined) out.default = property.default;
  if (property.options?.length) out.options = property.options;
  return out;
}

function serializeFilter(filter: LensFilter): Record<string, unknown> {
  if ("and" in filter) {
    return { and: filter.and.map(serializeFilter) };
  }
  if ("or" in filter) {
    return { or: filter.or.map(serializeFilter) };
  }
  if ("not" in filter) {
    return { not: serializeFilter(filter.not) };
  }
  const condition = filter as LensFilterCondition;
  const out: Record<string, unknown> = {
    property: condition.property,
    op: condition.op,
  };
  if (condition.value !== undefined) out.value = condition.value;
  return out;
}

function serializeSort(sort: LensSort): Record<string, unknown> {
  return { by: sort.by, dir: sort.dir };
}

function serializeView(view: LensView): Record<string, unknown> {
  const out: Record<string, unknown> = {
    id: view.id,
    name: view.name,
    type: view.type,
  };
  if (view.default) out.default = true;
  if (view.filters?.length) out.filters = view.filters.map(serializeFilter);
  if (view.sort?.length) out.sort = view.sort.map(serializeSort);
  if (view.limit != null) out.limit = view.limit;
  if (view.table) {
    const table: Record<string, unknown> = {
      columns: view.table.columns,
    };
    if (view.table.column_widths) table.column_widths = view.table.column_widths;
    if (view.table.row_height != null) table.row_height = view.table.row_height;
    out.table = table;
  }
  return out;
}

export function serializeLens(definition: LensDefinition): string {
  const out: Record<string, unknown> = {
    version: definition.version,
    name: definition.name,
  };
  if (definition.description) out.description = definition.description;

  out.dataset = {
    include: definition.dataset.include,
    ...(definition.dataset.exclude?.length
      ? { exclude: definition.dataset.exclude }
      : {}),
  };

  const props: Record<string, Record<string, unknown>> = {};
  for (const [id, prop] of Object.entries(definition.properties)) {
    props[id] = serializeProperty(prop);
  }
  out.properties = props;

  if (definition.filters?.length) {
    out.filters = definition.filters.map(serializeFilter);
  }
  if (definition.sort?.length) {
    out.sort = definition.sort.map(serializeSort);
  }

  out.views = definition.views.map(serializeView);

  if (definition.template) {
    const tmpl: Record<string, unknown> = {
      folder: definition.template.folder,
      filename: definition.template.filename,
    };
    if (definition.template.frontmatter) {
      tmpl.frontmatter = definition.template.frontmatter;
    }
    if (definition.template.body) tmpl.body = definition.template.body;
    out.template = tmpl;
  }

  return stringifyYaml(out, { lineWidth: 0 });
}

File: crates/cbase/src/serializer.rs
//! Serialize a definition back to `.cbase` YAML.
//!
//! The wire types are shaped to match the on-disk schema exactly, so a plain
//! YAML serialization round-trips through [`crate::parser::parse_cbase`]. This
//! is used to persist UI-driven changes (columns, sort, filters) to disk.

use crate::error::CbaseError;
use crate::types::CbaseDefinition;

/// Serialize a definition to `.cbase` YAML.
pub fn serialize_cbase(definition: &CbaseDefinition) -> Result<String, CbaseError> {
    serde_yaml::to_string(definition)
        .map_err(|e| CbaseError::Parse(format!("Failed to serialize .cbase: {e}")))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::parse_cbase;

    const MINIMAL_YAML: &str = "version: 1\nname: Test\ndataset:\n  include:\n    - \"**/*.md\"\nproperties:\n  title:\n    key: title\n    type: text\nviews:\n  - id: default\n    name: Default\n    type: table\n    table:\n      columns:\n        - title\n";

    #[test]
    fn roundtrips_minimal_definition() {
        let original = parse_cbase(MINIMAL_YAML, None).unwrap();
        let serialized = serialize_cbase(&original).unwrap();
        let reparsed = parse_cbase(&serialized, None).unwrap();

        assert_eq!(reparsed.name, original.name);
        assert_eq!(reparsed.dataset, original.dataset);
        assert_eq!(
            reparsed.properties.keys().collect::<Vec<_>>(),
            original.properties.keys().collect::<Vec<_>>()
        );
        assert_eq!(reparsed.views.len(), original.views.len());
        assert_eq!(
            reparsed.views[0].table.as_ref().map(|t| &t.columns),
            original.views[0].table.as_ref().map(|t| &t.columns)
        );
    }

    #[test]
    fn preserves_sort_and_filters() {
        let yaml = "version: 1\nname: With Sort\ndataset:\n  include:\n    - \"docs/**/*.md\"\n  exclude:\n    - \"docs/drafts/**\"\nproperties:\n  status:\n    key: status\n    type: select\n    options:\n      - open\n      - closed\n  priority:\n    key: priority\n    type: number\nfilters:\n  - property: status\n    op: \"!=\"\n    value: closed\nsort:\n  - by: priority\n    dir: desc\nviews:\n  - id: main\n    name: Main\n    type: table\n    default: true\n    table:\n      columns:\n        - status\n        - priority\n      column_widths:\n        status: 200\n";

        let original = parse_cbase(yaml, None).unwrap();
        let serialized = serialize_cbase(&original).unwrap();
        let reparsed = parse_cbase(&serialized, None).unwrap();

        assert_eq!(
            reparsed.dataset.exclude,
            Some(vec!["docs/drafts/**".to_string()])
        );
        assert_eq!(reparsed.filters.as_ref().map(|f| f.len()), Some(1));
        assert_eq!(reparsed.sort, original.sort);
        assert_eq!(reparsed.views[0].default, Some(true));
        assert_eq!(
            reparsed.views[0]
                .table
                .as_ref()
                .and_then(|t| t.column_widths.as_ref())
                .and_then(|w| w.get("status"))
                .copied(),
            Some(200.0)
        );
        assert_eq!(
            reparsed.properties["status"].options,
            Some(vec!["open".to_string(), "closed".to_string()])
        );
    }
}
